src/contracts/AbstractARM.sol

```diff
@@ -0,0 +1,652 @@
+    /// @param shares The amount of shares to burn
+    /// @return assets The amount of liquidity assets that would be received
+    function previewRedeem(uint256 shares) external view returns (uint256 assets) {
+        assets = convertToAssets(shares);
+    }
+
+    /// @notice Request to redeem liquidity provider shares for liquidity assets
+    /// @param shares The amount of shares the redeemer wants to burn for liquidity assets
+    /// @return requestId The index of the withdrawal request
+    /// @return assets The amount of liquidity assets that will be claimable by the redeemer
+    function requestRedeem(uint256 shares) external returns (uint256 requestId, uint256 assets) {
+        // Calculate the amount of assets to transfer to the redeemer
+        assets = convertToAssets(shares);
+
+        requestId = nextWithdrawalIndex;
+        uint120 queued = SafeCast.toUint120(withdrawsQueued + assets);
+        uint40 claimTimestamp = uint40(block.timestamp + CLAIM_DELAY);
+
+        // Store the next withdrawal request
+        nextWithdrawalIndex = SafeCast.toUint16(requestId + 1);
+        // Store the updated queued amount which reserves WETH in the withdrawal queue
+        withdrawsQueued = queued;
+        // Store requests
+        withdrawalRequests[requestId] = WithdrawalRequest({
+            withdrawer: msg.sender,
+            claimed: false,
+            claimTimestamp: claimTimestamp,
+            assets: SafeCast.toUint120(assets),
+            queued: queued
+        });
+
+        // burn redeemer's shares
+        _burn(msg.sender, shares);
+
+        // Remove the redeemed assets from the last available assets
+        lastAvailableAssets -= SafeCast.toInt128(SafeCast.toInt256(assets));
+
+        emit RedeemRequested(msg.sender, requestId, assets, queued, claimTimestamp);
+    }
+
+    /// @notice Claim liquidity assets from a previous withdrawal request after the claim delay has passed
+    /// @param requestId The index of the withdrawal request
+    /// @return assets The amount of liquidity assets that were transferred to the redeemer
+    function claimRedeem(uint256 requestId) external returns (uint256 assets) {
+        // Load the structs from storage into memory
+        WithdrawalRequest memory request = withdrawalRequests[requestId];
+
+        require(request.claimTimestamp <= block.timestamp, "Claim delay not met");
+        // Is there enough liquidity to claim this request?
+        require(
+            request.queued <= withdrawsClaimed + IERC20(liquidityAsset).balanceOf(address(this)),
+            "Queue pending liquidity"
+        );
+        require(request.withdrawer == msg.sender, "Not requester");
+        require(request.claimed == false, "Already claimed");
+
+        // Store the request as claimed
+        withdrawalRequests[requestId].claimed = true;
+        // Store the updated claimed amount
+        withdrawsClaimed += request.assets;
+
+        assets = request.assets;
```

Comment:
⚪️ I think the `assets = request.assets;` line should go above the `withdrawsClaimed += request.assets;` line, and then that should change to `withdrawsClaimed += assets`; This way the storage set, the event, and the transfer are all working off the same value. Saves a warm gas read too.

While that does split up the storage writes from being adjacent to each other, I think it's worth it in this case.

Reply:
Nice. done